pkg/bloomcompactor/bloomcompactor.go

```diff
@@ -158,33 +165,256 @@ func New(cfg Config,
+	// Keep track of tenants owned by this shard, so that we can delete the local files for all other users.
+	errs := multierror.New()
+	ownedTenants := make(map[string]struct{}, len(tenants))
+	for _, tenant := range tenants {
+		logger := log.With(c.logger, "tenant", tenant)
+
+		// Ensure the context has not been canceled (ie. compactor shutdown has been triggered).
+		if err := ctx.Err(); err != nil {
+			return fmt.Errorf("interrupting compaction of tenants: %w", err)
+		}
+
+		// Ensure the tenant ID belongs to our shard.
+		if !c.sharding.OwnsTenant(tenant) {
+			c.metrics.compactionRunSkippedTenants.Inc()
+			level.Debug(logger).Log("msg", "skipping tenant because it is not owned by this shard")
+			continue
+		}
+
+		ownedTenants[tenant] = struct{}{}
+
+		if err := c.compactTenantWithRetries(ctx, sc, tableName, tenant); err != nil {
+			switch {
+			case errors.Is(err, context.Canceled):
+				// We don't want to count shutdowns as failed compactions because we will pick up with the rest of the compaction after the restart.
+				level.Info(logger).Log("msg", "compaction for tenant was interrupted by a shutdown")
+				return nil
+			default:
+				c.metrics.compactionRunFailedTenants.Inc()
+				level.Error(logger).Log("msg", "failed to compact tenant", "err", err)
+				errs.Add(err)
+			}
+			continue
+		}
+
+		c.metrics.compactionRunSucceededTenants.Inc()
+		level.Info(logger).Log("msg", "successfully compacted tenant")
+	}
+
+	return errs.Err()
+
+	// TODO: Delete local files for unowned tenants, if there are any.
+}
+
+func (c *Compactor) compactTenant(ctx context.Context, sc storeClient, tableName string, tenant string) error {
+	level.Info(c.logger).Log("msg", "starting compaction of tenant", "tenant", tenant)
+
+	// Ensure the context has not been canceled (ie. compactor shutdown has been triggered).
+	if err := ctx.Err(); err != nil {
+		return err
+	}
+
+	// Tokenizer is not thread-safe so we need one per goroutine.
+	bt, _ := v1.NewBloomTokenizer(prometheus.DefaultRegisterer)
+
+	// TODO: Use ForEachConcurrent?
+	errs := multierror.New()
+	if err := sc.indexShipper.ForEach(ctx, tableName, tenant, func(isMultiTenantIndex bool, idx shipperindex.Index) error {
+		if isMultiTenantIndex {
+			return fmt.Errorf("unexpected multi-tenant")
+		}
+
+		// TODO: Make these casts safely
+		if err := idx.(*tsdb.TSDBFile).Index.(*tsdb.TSDBIndex).ForSeries(
+			ctx, nil,
+			0, math.MaxInt64, // TODO: Replace with MaxLookBackPeriod
+			func(labels labels.Labels, fingerprint model.Fingerprint, chksMetas []tsdbindex.ChunkMeta) {
+				job := NewJob(tenant, tableName, idx.Path(), fingerprint, labels, chksMetas)
+
+				ownsJob, err := c.sharding.OwnsJob(job)
+				if err != nil {
+					c.metrics.compactionRunSkippedJobs.Inc()
+					level.Error(c.logger).Log("msg", "failed to check if compactor owns job", "job", job, "err", err)
+					errs.Add(err)
+					return
+				}
+				if !ownsJob {
+					c.metrics.compactionRunSkippedJobs.Inc()
```

Comment:
I can see this metric name `compactionRunSkippedJobs` can get confusing in the future because there will be whole compactions skipped because there is no change in indexes and chunks that are already compacted. 
why do we want to count jobs not owned by the shard? we must assume it will be processed by another shard.

 I can see it as a back up to see if we are missing a subgroup of jobs all together, maybe we can use a metric like jobNotOwned, still it doesn't help us find which jobs.

Reply:
> because there will be whole compactions skipped because there is no change in indexes

What about renaming the metric to `unowned_jobs` instead?

> why do we want to count jobs not owned by the shard? we must assume it will be processed by another shard.

I think this will be useful for debugging proposes. We can look at the total number started compaction jobs and the number of finished jobs. ideally the number of started should be the same as finished. If they differ we can use the unowned metric to see if there are jobs not being owned by any compactor (if the sum of unowned jobs. more than jobs started).

Having said that, we'd need to add the `started` metric...

Reply:
yes i agree that it's good for debugging purposes. `unowned_jobs` sounds the most appropriate.

Reply:
Done :)